### PR TITLE
Adding slightly fancier XP bar

### PR DESCRIPTION
**Pull request by thebigMuh**
- PCX background
- end pixel fades in gradually
- shows detailed info on mouse over
image
- gold bar + gold text for max level
image

### CODE REVIEW

File: SourceX/qol.cpp [hunk 1]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
-	int numDividers = 10;
-	int barColor = 198;
-	int emptyBarColor = 0;
-	int frameColor = 196;
-	bool space = true; // add 1 pixel separator on top/bottom of the bar
-
-	PrintGameStr(out, xPos - 22, yPos + 6, "XP", COL_WHITE);
-	int charLevel = plr[myplr]._pLevel;
-	if (charLevel == MAXCHARLEVEL - 1)
+	const PlayerStruct& player = plr[myplr];
```
Comment (Iniquitatis) [inline SourceX/qol.cpp:193]: Finally! I was hoping that someone will start replacing all these `plr[whatever]` with references.

File: SourceX/qol.cpp [hunk 2]
```diff
@@ -10,10 +10,25 @@
...
 namespace {
 
+namespace xpbar {
+	constexpr int barWidth = 307;
+	constexpr int barHeight = 5;
+
+	constexpr int goldGradient[] = { 0xCF, 0xCE, 0xCD, 0xCC, 0xCB, 0xCA, 0xC9, 0xC8, 0xC7, 0xC6, 0xC5, 0xC4, 0xC3, 0xC2, 0xC1, 0xC0 };
+	constexpr int goldGrades = 12;
+
+	constexpr int whiteGradient[] = { 0xFE, 0xFD, 0xFC, 0xFB, 0xFA, 0xF9, 0xF8, 0xF7, 0xF6, 0xF5, 0xF4, 0xF3, 0xF2, 0xF1, 0xF0, 0xFF };
```
Comment (Iniquitatis) [inline SourceX/qol.cpp:20]: `[0] = 0xFE`, `[15] = 0xFF` — is this intentional? `goldGradient` has a different order.

Comment (thebigMuh) [reply SourceX/qol.cpp:20]: Yes, intentional. The black/white gradient has unexpected order.

Compare the top row (greyscale gradient) in this screenshot to the other 16 color gradients (4th row is the gold gradient):
image

File: SourceX/qol.cpp (hunk 1 above)
Comment (AJenbo) [reply SourceX/qol.cpp:193]: Slowly but surely it's happening.

Comment (julealgon) [reply SourceX/qol.cpp:193]: > Finally! I was hoping that someone will start replacing all these plr[whatever] with references.

You are a few months late it seems @Iniquitatis 🤣

**Review by julealgon — CHANGES_REQUESTED** [review]
Please rebase your changes.

Also, please consider splitting the 2 things you are doing (updating the graphics, and adding the text information) in at least 2 commits: one for each.

File: SourceX/qol.cpp [hunk 3]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
-	if (plr[myplr]._pExperience < prevXp)
+	const int prevXp = ExpLvlsTbl[charLevel - 1];
+	if (player._pExperience < prevXp)
 		return;
 
-	Uint64 prevXpDelta_1 = plr[myplr]._pExperience - prevXp;
-	int prevXpDelta = ExpLvlsTbl[charLevel] - prevXp;
-	int visibleBar = barWidth * prevXpDelta_1 / prevXpDelta;
+	Uint64 prevXpDelta_1 = player._pExperience - prevXp;
+	Uint64 prevXpDelta = ExpLvlsTbl[charLevel] - prevXp;
```
Comment (julealgon) [inline SourceX/qol.cpp:219]: Please avoid touching unrelated code. Uin64 is actually not the appropriate type to use here.

**Review by julealgon — CHANGES_REQUESTED** [review]

File: SourceX/qol.cpp [hunk 4]
```diff
@@ -10,10 +10,25 @@
 namespace devilution {
 namespace {
 
+namespace xpbar {
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:13]: It makes the xpbar namespace private to the compilation unit. So I can have my cake (not exposing xpbar to the outside world or causing collisions) and eat it too (having the static data in a namespace).

File: SourceX/qol.cpp [hunk 5]
```diff
@@ -10,10 +10,25 @@
 namespace devilution {
 namespace {
 
+namespace xpbar {
+	constexpr int barWidth = 307;
+	constexpr int barHeight = 5;
+
+	constexpr int goldGradient[] = { 0xCF, 0xCE, 0xCD, 0xCC, 0xCB, 0xCA, 0xC9, 0xC8, 0xC7, 0xC6, 0xC5, 0xC4, 0xC3, 0xC2, 0xC1, 0xC0 };
+	constexpr int goldGrades = 12;
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:18]: There are about a dozen ways to skin this cat.

I don't really want to change the simplicity of the source code just to get some fancy container in there. The gradient array is actually longer than the number of colors I'm using (otherwise it would have been too yellow/too bright), that's why I have an explicit second variable that limits it to the darker 12 colors.

File: SourceX/qol.cpp [hunk 6]
```diff
@@ -27,6 +42,18 @@ int GetTextWidth(const char *s)
...
+// Returns pointer to first char after the written number.
+char* PrintWithSeparator(char* out, long long n)
+{
+	if (n < 1000) {
+		return out + sprintf(out, "%lld", n);
+	}
+
+	char* append = PrintWithSeparator(out, n / 1000);
+	return append + sprintf(append, ",%03lld", n % 1000);
+}
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:55]: I could come up with arbitrarily fancy ways to do this, but ultimately it just prints a number into a buffer, without allocating memory (unless sprintf is funny).

This is about as simple and straightforward as it gets. There is a POSIX 2008 extension to printf style format strings that allows using the ' character for decimal grouping using locale specific rules, but I really doubt that support for this can be assumed on all platforms.

File: SourceX/qol.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:193]: Can't see anything in the coding style guide about this. Ultimately, as long as only a single variable per line is declared (which is generally regarded as a good base rule), it doesn't matter.

File: SourceX/qol.cpp [hunk 7]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
+
+	if (MouseX < backX || MouseX >= backX + xpbar::backWidth ||
+		MouseY < backY || MouseY >= backY + xpbar::backHeight)
+		return false;
+
+	const PlayerStruct& player = plr[myplr];
+
+	const int charLevel = player._pLevel;
+
+	char tempstr[64];
```
[the remark this answers is not in the document]
Comment (qndel) [reply SourceX/qol.cpp:255]: Why not? 🤪

File: SourceX/qol.cpp [hunk 8]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
+	FastDrawHorizLine(out, xPos, yPos + 2, fullBar, xpbar::whiteGradient[xpbar::whiteGrades - 1]);
+	FastDrawHorizLine(out, xPos, yPos + 3, fullBar, xpbar::whiteGradient[xpbar::whiteGrades / 2 - 1]);
+
+	// End pixels appear gradually
+	SetPixel(out, xPos + fullBar, yPos + 1, xpbar::whiteGradient[fade * 3 / 4]);
+	SetPixel(out, xPos + fullBar, yPos + 2, xpbar::whiteGradient[fade]);
+	SetPixel(out, xPos + fullBar, yPos + 3, xpbar::whiteGradient[fade / 2]);
+}
+
+bool CheckXPBarInfo(void)
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:239]: Old habit... makes plain text searching for declaration vs. invocation simpler.
I can remove it if it's an issue.

File: SourceX/qol.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (qndel) [reply SourceX/qol.cpp:193]: Yes but it will cause differences when autoformatted (at least my VS2019 autoformats & to the right side and I'm assuming it's doing what devilutionx is telling it to

[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:193]: I'm using the clang-tidy that came with the project (at least I think I am), but I don't think the rules in there are quite correct. It for example complains about variable names, since the rules state that all variable names have to be exclusively lower case.

[the remark this answers is not in the document]
Comment (qndel) [reply SourceX/qol.cpp:193]: I guess we have to ask @AJenbo

File: SourceX/qol.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:255]: It should be long enough for all reasonable values.
Also infostr[64] and panelstr[4][64] exist and kind of set the limit for what is feasibly printed into the info box.
Replacing all of that with some constant is I guess outside the scope of this PR.

File: SourceX/qol.cpp (hunk 3 above)
Comment (thebigMuh) [reply SourceX/qol.cpp:219]: The calculations I'm doing exceed SInt32 range. Rather than casting all the time, I'm using a sufficiently sized datatype for all involved variables.

File: SourceX/qol.cpp [hunk 9]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
-	int prevXp = ExpLvlsTbl[charLevel - 1];
-	if (plr[myplr]._pExperience < prevXp)
+	const int prevXp = ExpLvlsTbl[charLevel - 1];
+	if (player._pExperience < prevXp)
 		return;
 
-	Uint64 prevXpDelta_1 = plr[myplr]._pExperience - prevXp;
-	int prevXpDelta = ExpLvlsTbl[charLevel] - prevXp;
-	int visibleBar = barWidth * prevXpDelta_1 / prevXpDelta;
+	Uint64 prevXpDelta_1 = player._pExperience - prevXp;
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:218]: See below.

File: SourceX/qol.cpp [hunk 10]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
+	SetPixel(out, xPos + fullBar, yPos + 3, xpbar::whiteGradient[fade / 2]);
+}
+
+bool CheckXPBarInfo(void)
+{
+	if (!sgOptions.Gameplay.bExperienceBar)
+		return false;
+
+	const int backX = (gnScreenWidth / 2) - 155;
+	const int backY = gnScreenHeight - 11;
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:245]: It means 155 pixels to the left and 11 pixels up.
Not sure that introducing yet another set of constants would help rather than hinder readability.

File: SourceX/qol.cpp [hunk 11]
```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
...
-	FastDrawHorizLine(out, xPos - 1, yPos + barHeight, barWidth + 2, frameColor);
-	FastDrawVertLine(out, xPos - 1, yPos - 1, barHeight + 2, frameColor);
-	FastDrawVertLine(out, xPos + barWidth, yPos - 1, barHeight + 2, frameColor);
-	for (int i = 1; i < numDividers; i++)
-		FastDrawVertLine(out, xPos - 1 + (barWidth * i / numDividers), yPos - dividerHeight + 3, barHeight, 245);
+	sprintf(PrintWithSeparator(tempstr, ExpLvlsTbl[charLevel] - player._pExperience), " to Level %d", charLevel + 1);
+	AddPanelString(tempstr, true);
 
-	FillRect(out, xPos, yPos + (space ? 1 : 0), visibleBar, barHeight - (space ? 2 : 0), barColor);
+	return true;
```
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:289]: I don't know why the diff felt the need to mix the xp bar drawing code into the xp info printing code. Diff tools can be like that sometimes.

I think the functions are short enough that they can be reviewed on their own, however, rather than a line by line comparison with the previous code.

File: SourceX/qol.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AJenbo) [reply SourceX/qol.cpp:193]: clang-format is for style.

clang-tidy is for refactoring and has not been applied or adjusted yet.

File: SourceX/qol.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (AJenbo) [reply SourceX/qol.cpp:245]: there are already macros/defines for placing content on the panel, use them instead of screen space.

File: SourceX/qol.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (AJenbo) [reply SourceX/qol.cpp:239]: would prefer for consistancy

File: SourceX/qol.cpp (hunk 3 above)
Comment (qndel) [reply SourceX/qol.cpp:219]: Subtracting 2 positive SInt32 doesn't need 64 :D

Comment (thebigMuh) [reply SourceX/qol.cpp:219]: Nope, but xpbar::barWidth * prevXpDelta_1 / prevXpDelta (right below) does. If all participants are integers, you get interesting results.

So either you cast (yuck) or you just make the types 64 bit wide by default, which seemed fine and less contentious. Guess I was wrong :D!

File: SourceX/qol.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp:193]: Ah! Will investigate how to teach VS that...

[the remark this answers is not in the document]
Comment (qndel) [reply SourceX/qol.cpp:193]: I copy/paste my code 😅

File: SourceX/qol.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (julealgon) [reply SourceX/qol.cpp:55]: > This is about as simple and straightforward as it gets.

I'm sorry but I don't believe you, even not knowing much about C++ at all. There must be a way to do something like `Print(number, "0.000.000")` or an equivalent thing.

File: SourceX/qol.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (julealgon) [reply SourceX/qol.cpp:255]: > Replacing all of that with some constant is I guess outside the scope of this PR

Would be a nice preliminary refactor, though. By introducing extra things that depend on the magic number, you are making an existing technical debt a bit worse than before.

**_make the change easy, then make the easy change_**

File: SourceX/qol.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp]: https://stackoverflow.com/questions/43482488/how-to-format-a-number-with-thousands-separator-in-c-c

Knock yourself out :D

[the remark this answers is not in the document]
Comment (thebigMuh) [reply SourceX/qol.cpp]: Btw. of all the answers, the most C++ way is probably the locale one.

For someone reading the code, not only does it require that the person has an idea what locales are and how they work, but it also expects to stream the number into a destination - usually something like a stringstream.

If I want to get away without the additional memory allocations this will incur, I have to additionally derive my own memory stream from std::streambuf and use that.

Instead of that, most likely I could throw out the entire char buffer thing, and instead use a std::string, print into that and then fetch the result and throw it into PrintPanelStr, which also does memory allocations along the way...

All of which is massively more convoluted than just printing with that simple function.

File: Source/qol.cpp [hunk 12]
```diff
@@ -12,10 +12,25 @@
 namespace devilution {
 namespace {
 
+namespace xpbar {
```
Comment (glebm) [inline Source/qol.cpp:15]: I think it'd be cleaner to move the xpbar stuff to its own file. This one is growing unwieldy.

File: Source/qol.cpp [hunk 13]
```diff
@@ -12,10 +12,25 @@
 namespace devilution {
 namespace {
 
+namespace xpbar {
+constexpr int barWidth = 307;
+constexpr int barHeight = 5;
+
+constexpr int goldGradient[] = { 0xCF, 0xCE, 0xCD, 0xCC, 0xCB, 0xCA, 0xC9, 0xC8, 0xC7, 0xC6, 0xC5, 0xC4 };
+constexpr int goldGrades = SDL_arraysize(goldGradient);
```
Comment (glebm) [inline Source/qol.cpp:20]: Optional suggestion (in case you think it's better):

```suggestion
constexpr std::array<std::uint8_t, 12> = { 0xCF, 0xCE, 0xCD, 0xCC, 0xCB, 0xCA, 0xC9, 0xC8, 0xC7, 0xC6, 0xC5, 0xC4 };
```

File: Source/qol.cpp (hunk 12 above)
Comment (glebm) [reply Source/qol.cpp:15]: Perhaps `Source/qol/xpbar.cpp` /cc @AJenbo

Comment (julealgon) [reply Source/qol.cpp:15]: I see no reason for the "qol" in the name. It is already under a toggle, the file itself doesn't need the extra info.

Comment (thebigMuh) [reply Source/qol.cpp:15]: All for splitting the qol stuff up, but to be honest this would inflate the PR even further. So I'd suggest doing this in a separate commit.

File: Source/qol.cpp (hunk 13 above)
Comment (thebigMuh) [reply Source/qol.cpp:20]: Sure, if it's OK that I drag the STL into this :)

File: Source/qol.cpp (hunk 12 above)
Comment (AJenbo) [reply Source/qol.cpp:17]: > Perhaps `Source/qol/xpbar.cpp` /cc @AJenbo

I was just thinking of that earlier today.

Comment (AJenbo) [reply Source/qol.cpp:17]: > All for splitting the qol stuff up, but to be honest this would inflate the PR even further. So I'd suggest doing this in a separate commit.

I think that is exactly what is being suggestd.

File: SourceX/qol.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (glebm) [reply SourceX/qol.cpp]: > I'm sorry but I don't believe you

The way it's done currently is indeed the way to do it from C++11 to C++20 (when not using any support libraries, like `fmt` or Abseil)

There are alternatives but they're bad for other reasons.

File: SourceX/qol.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (glebm) [reply SourceX/qol.cpp]: https://devblogs.microsoft.com/cppblog/clangformat-support-in-visual-studio-2017-15-7-preview-1/

**Comment by glebm** [comment]
Resolved a bunch of comments that have already been addressed (comment authors: please click resolve once it's been addressed)

**Comment by thebigMuh** [comment]
Closing this as it was pretty messed up by my git shenanigans.

Opened new PR here: https://github.com/diasurgical/devilutionX/pull/1597
